### PR TITLE
test: tweak up test-utils

### PR DESCRIPTION
**Pull request by antongolub**
Refactor test-utils:
* Introduce `test`, `test.only` and  `test.skip` helpers
* Highlight failed tasks in red
* Apply CLI filter argument to test file paths
* Add test exec timeout

- [x] Tests pass

### CODE REVIEW

**Comment by antongolub** [comment]
@antonmedv , rfr

**Comment by antonmedv** [comment]
Maybe it’s already a time to switch to something like ava.li?

**Comment by antongolub** [comment]
Right now I'd like to have more control over the test execution logic. In the future, yes, we will migrate to smth else, so I made the contract compatible with basic mocha/jest/ava/uvu test api.

**Comment by antongolub** [comment]
I really like our custom <filter any> cli api. Just try. This is super handy for developing in debug mode. It might even be worth publishing our test snippet as a package.

**Comment by antongolub** [comment]
@antonmedv,

let's move this forward.

**Comment by antonmedv** [comment]
Sure. Will do it today

File: test/test-utils.mjs [hunk 1]
```diff
@@ -13,24 +13,96 @@
...
+export const test = async function (name, cb, focus, skip) {
+  const filter = RegExp(process.argv[3] || '.')
+  const {group, meta} = this
+  const file = meta ? relative(process.cwd(), fileURLToPath(meta.url)) : ''
+
+  if (filter.test(name) || filter.test(group) || filter.test(file)) {
+    focused += +!!focus
+    queued++
+
+    await warmup
```
Comment (antonmedv) [inline test/test-utils.mjs:65]: Why this warmup needed?

Comment (antongolub) [reply test/test-utils.mjs:65]: There's no test files loader (like in jest), so we don't know how many files and how many test directives (only, skip) there will be at all. `ava/uvu` uses `test.run()` for this purpose. I think a short delay is better for our case.

Comment (antonmedv) [reply test/test-utils.mjs:65]: I’m still not getting it. Logic with timeouts seems over complicated. Maybe remove this part and don’t introduce timeouts for tests?

Comment (antongolub) [reply test/test-utils.mjs:65]: Here's an example:
```js
// some.test.js
test(...)
test(...)
// one-more.test.js
test(...)
test.only()
```
We need to check, that there's no `only` in the entire test suite _before_ running the first test.

Comment (antonmedv) [reply test/test-utils.mjs:65]: I see. Isn’t it too much? I have doubts about implementing complicated testing framework.

Comment (antongolub) [reply test/test-utils.mjs:65]: Race here: resolve test callback promise or throw timeout exception.
```js
const timeout = (promise, ms, exception = `TimeoutException: exec time exceeds ${ms}ms`) => {
  let timer
  return Promise.race([
    promise,
    new Promise((_r, rej) => timer = setTimeout(rej, ms, exception))
  ]).finally(() => clearTimeout(timer))
}
```

Comment (antonmedv) [reply test/test-utils.mjs:65]: I see. What test use this timeout functionality?

Comment (antongolub) [reply test/test-utils.mjs:65]: All of them: 
```js
const run = singleThread((cb) => timeout(cb(), 5000)).
```
Default timeout for every test. Like in jest.

Comment (antonmedv) [reply test/test-utils.mjs:65]: But I why do we need timeouts in a first place?

Comment (antongolub) [reply test/test-utils.mjs:65]: Just to highlight problems: If some test exceeds 5000 ms, it may indicate an implementation issue.

Comment (antonmedv) [reply test/test-utils.mjs:65]: Try, this will lead to "stuck" GitHub action, as we saw =) Let's discuss this tomorrow?

Comment (antongolub) [reply test/test-utils.mjs:65]: Deal. I'm also available for tg / zoom / emails after 18:00 (UTC +3).

Comment (antongolub) [reply test/test-utils.mjs:65]: * I've introduced two levels of protection agaist hanging: this one race-wrapped timeout and `- run: timeout 30 npm test` inside the test action.
* I can drop `warmup` if test src loading goes to test-utils too

Comment (antonmedv) [reply test/test-utils.mjs:65]: I still not convinced we need Promise.race stuff. Can we solve it via GitHub actions timeout?

Comment (antongolub) [reply test/test-utils.mjs:57]: Ok, removed this part.

Comment (antonmedv) [reply test/test-utils.mjs:57]: Looks like still here. Also, is timeout available on windows? May switch to GH action one?

Comment (antongolub) [reply test/test-utils.mjs:57]: * The diff: https://github.com/google/zx/pull/360/commits/220c8a6522714c2d3ed3d45af1a6b7a7551bbc49
* `timeout` for Windows:
  * https://superuser.com/questions/520527/windows-command-line-time-limit
  * https://stackoverflow.com/questions/46374893/timeout-equivalent-for-windows
  
I found [`timeout-cli`](https://github.com/anaisbetts/timeout-cli) pkg, but it's definitely overcomlicated inside. How about just `npx zx --experimental <<< 'withTimeout(30000)`npm run test`'`?

Comment (antongolub) [reply test/test-utils.mjs:57]: I can add windows to the test matrix later.

Comment (antonmedv) [reply test/test-utils.mjs:57]: Why not just use https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idtimeout-minutes ?

Comment (antongolub) [reply test/test-utils.mjs:57]: Finally pushed.

Comment (antongolub) [reply test/test-utils.mjs:57]: btw, I've landed this `test-utils.js` snippet to [my pet project](https://github.com/antongolub/npm-registry-firewall/blob/master/src/test/test-utils.js), it works fine.

**Review by antonmedv — APPROVED** [review]

**Pull request merged**